src/core/toga/events.py

```diff
@@ -0,0 +1,190 @@
+
+    def raise_event(self, event, *args, **kwargs):
+        """Call the callback associated with an event, if any
+
+        Args:
+            event (str): The name of the event to raise callbacks for
+
+        All other positional and keyword arguments are passed to the callback
+        function. If a bad event string is passed, UndefinedEventRaised
+        exception is raised.
+        """
+        descr = self._get_event_descriptor(event)
+        if descr is None:
+            raise UndefinedEventRaised(event)
+        descr.raise_event(self, *args, **kwargs)
+
+    @classmethod
+    def _get_event_descriptor(cls, event):
+        # Its unfortunate, but it seems there is no method or function that is
+        # internal to the interpreter and that we could use to scan the MRO
+        # and access the descriptor object without also masking it away and
+        # implicitly calling its __get__ or __set__ methods.
+        for klass in cls.mro():
+            descriptor = klass.__dict__.get(event)
+            if descriptor is None:
+                continue
+            if isinstance(descriptor, Event):
+                return descriptor
+            else:
+                return None
+        else:
+            return None
+
+
+class Event:
+    """Descriptor class for event callback attributes
+
+    Args:
+        doc (str): Documentation string for the event
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def __repr__(self):
+        return '{} Event'.format(self._name or "Unnamed")
+
+    def __set_name__(self, owner, name):
+        self._name = name
+        self._attr_name = '_{}'.format(name)
+
+    def __get__(self, instance, owner):
+        if instance is None:
+            return None
+        return getattr(instance, self._attr_name, None)
+
+    def __set__(self, instance, callback):
+        setattr(instance, self._attr_name, callback)
+        self._notify_impl(instance, when_none=True)
+
+    def _notify_impl(self, instance, when_none=False):
+        """Notifies the instance backend implementation object when a new
+        callback is set if that object implements the `set_event_handler`
+        method
+        """
+        callback = self.__get__(instance, type(instance))
+        if callback is None and not when_none:
+            return
+        impl = getattr(instance, '_impl', None)
+        if impl is not None and hasattr(impl, 'set_event_handler'):
+            impl.set_event_handler(self._name, callback)
+
+    def raise_event(self, instance, *args, **kwargs):
+        """Call the callback associated with the event, if any
+
+        Args:
+            instance (:obj:`EventSource`): A widget to call the event for
+
+        All other arguments are passed on the the callback
+        """
+        callback = self.__get__(instance, type(instance))
+        if callback is None:
+            return
+        wrapped_handler(instance, callback)(instance, *args, **kwargs)
```

Comment:
If I'm reading this right, this will result in the handler being wrapped on every event raised, correct? Should this be done in the `__set__` method?

Reply:
Well I wanted to preserve the following idiom WRT callback attributes:

when o.attr = x then x == o.attr

So therefore I preserve the unwrapped callback value. I could also pre-cache the wrapped handler, but that seemed like premature optimization to me. LMK if you disagree.

Reply:
I can see how maintaining the direct comparison could be useful; however, I'm not sure I agree it's premature optimization to pre-compute and cache the wrapper handler. Responsiveness of events has a direct consequence in perceived "snappiness" of the user interface, so pre-computing anything we can would seem beneficial.

Reply:
Alright, added wrapped callback pre-caching, as well as MRO search results caching so that raising event would potentially be faster.